File: VB/CSPDashboardExample/Models/DashboardModel.vb
Imports System.Security.Cryptography
Imports Microsoft.AspNetCore.Mvc
Imports Microsoft.AspNetCore.Mvc.RazorPages

Namespace CSPDashboardExample.Models

    Public Class DashboardModel
        Inherits PageModel

        Public Property Nonce As String

        Public Sub New()
            Dim nonceBytes = New Byte(31) {}
            Dim generator = RandomNumberGenerator.Create()
            generator.GetBytes(nonceBytes)
            Nonce = Convert.ToBase64String(nonceBytes)
        End Sub

        Public Function OnGet() As IActionResult
            HttpContext.Response.Headers.Add("Content-Security-Policy", "img-src data: https: http:;" & String.Format("script-src 'self' 'nonce-{0}';", Nonce) + String.Format("style-src 'self' 'nonce-{0}';", Nonce))
            Return Page()
        End Function
    End Class
End Namespace

File: CS/CSPDashboardExample/Models/DashboardModel.cs
using System.Security.Cryptography;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.RazorPages;

namespace CSPDashboardExample.Models
{
    public class DashboardModel: PageModel
    {
        public string Nonce { get; set; }

        public DashboardModel() {
            var nonceBytes = new byte[32];
            var generator = RandomNumberGenerator.Create();
            generator.GetBytes(nonceBytes);
            Nonce = Convert.ToBase64String(nonceBytes);
        }

        public IActionResult OnGet() {
            HttpContext.Response.Headers.Add("Content-Security-Policy",
                "img-src data: https: http:;" +
                string.Format("script-src 'self' 'nonce-{0}';", Nonce) +
                string.Format("style-src 'self' 'nonce-{0}';", Nonce) 
                    );
            return Page();
        }
    }
}
